### PR TITLE
Feature/3099 nan tolerance

### PR DESCRIPTION
**Pull request by piotrb5e3**
Hi,

This is a proposed solution to #3099.

I unified the code of all tolerance matchers and improved handling around NaNs.

It has one issue: since the new code operates all on doubles, the old "Match close enough numbers" test fails, as the ranges of acceptable values are slightly different with double rounding. Not really sure what to do about it.

I'd also happily remove `FloatToleranceMatcher`, but I'm not sure if it's not part of the API.

### CODE REVIEW

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/numeric/ToleranceMatcher.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package io.kotest.matchers.numeric
+
+import io.kotest.matchers.Matcher
+import io.kotest.matchers.MatcherResult
+import kotlin.math.abs
+
+open class ToleranceMatcher<T : Number>(private val expected: T, private val tolerance: Double) : Matcher<T> {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/numeric/ToleranceMatcher.kt:7]: Interesting solution by using `Number`!

**Review by sksamuel — APPROVED** [review]

**Comment by sksamuel** [comment]
Just need to fix the test @piotrb5e3 and I think we're good!

**Comment by Kantis** [comment]
In my opinion we should revert the unification of ToleranceMatchers. It'll lead to some duplicate code, but it will let us be as correct as possible which I think is more important.

**Comment by sksamuel** [comment]
I agree

**Comment by LeoColman** [comment]
@piotrb5e3 Could you give maintainers access to your branch so we can fix the git history?

**Comment by piotrb5e3** [comment]
@LeoColman Should I add specific people to the access list for this branch? _Allow edits by maintainers_ is already enabled.

**Pull request closed without merging**